sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```
@@ -0,0 +1,110 @@
+
+
+/**
+ * Generates a [[Projection]] that returns an [[UnsafeRow]].
+ *
+ * Note: The returned UnsafeRow will be pointed to a scratch buffer inside the projection.
+ */
+object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafeProjection] {
+
+  protected def canonicalize(in: Seq[Expression]): Seq[Expression] =
+    in.map(ExpressionCanonicalizer.execute)
+
+  protected def bind(in: Seq[Expression], inputSchema: Seq[Attribute]): Seq[Expression] =
+    in.map(BindReferences.bindReference(_, inputSchema))
+
+  protected def create(expressions: Seq[Expression]): UnsafeProjection = {
+    val ctx = newCodeGenContext()
+    val codes = expressions.map(_.gen(ctx))
+    val all_exprs = codes.map(_.code).mkString("\n")
+    val fixedSize = 8 * codes.length + UnsafeRow.calculateBitSetWidthInBytes(codes.length)
+    val stringWriter = "org.apache.spark.sql.catalyst.expressions.StringUnsafeColumnWriter"
+    val binaryWriter = "org.apache.spark.sql.catalyst.expressions.BinaryUnsafeColumnWriter"
+    val additionalSize = expressions.zipWithIndex.map { case (e, i) =>
+      e.dataType match {
+        case StringType =>
+          s" + (${codes(i).isNull} ? 0 : $stringWriter.getSize(${codes(i).primitive}))"
+        case BinaryType =>
+          s" + (${codes(i).isNull} ? 0 : $binaryWriter.getSize(${codes(i).primitive}))"
+        case _ => ""
+      }
+    }.mkString("")
+
+    val writers = expressions.zipWithIndex.map { case (e, i) =>
+      val update = e.dataType match {
+        case dt if ctx.isPrimitiveType(dt) =>
+          s"${ctx.setColumn("target", dt, i, codes(i).primitive)}"
+        case StringType =>
+          s"cursor += $stringWriter.write(target, ${codes(i).primitive}, $i, cursor)"
+        case BinaryType =>
+          s"cursor += $binaryWriter.write(target, ${codes(i).primitive}, $i, cursor)"
+        case _ =>
+          throw new Exception(s"Not supported DataType: ${e.dataType}")
```

need to support nulltype here.

BTW let's throw UnsupportedOperationException rather than Exception.